os/fs/procfs/fs_procfslogsave.c

```diff
@@ -0,0 +1,191 @@
+        unsigned int linesize;          /* Number of valid characters in line[] */
+        char line[10];     /* Pre-allocated buffer for formatted lines */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+/* File system methods */
+
+static int logsave_open(FAR struct file *filep, FAR const char *relpath, int oflags, mode_t mode);
+static int logsave_close(FAR struct file *filep);
+static ssize_t logsave_read(FAR struct file *filep, FAR char *buffer, size_t buflen);
+static int logsave_dup(FAR const struct file *oldp, FAR struct file *newp);
+static int logsave_stat(FAR const char *relpath, FAR struct stat *buf);
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Variables
+ ****************************************************************************/
+
+/* See fs_mount.c -- this structure is explicitly externed there.
+ * We use the old-fashioned kind of initializers so that this will compile
+ * with any compiler.
+ */
+
+const struct procfs_operations logsave_operations = {
+	logsave_open,				/* open */
+	logsave_close,				/* close */
+	logsave_read,				/* read */
+	NULL,						/* write */
+
+	logsave_dup,				/* dup */
+
+	NULL,						/* opendir */
+	NULL,						/* closedir */
+	NULL,						/* readdir */
+	NULL,						/* rewinddir */
+
+	logsave_stat				/* stat */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: logsave_open
+ ****************************************************************************/
+
+static int logsave_open(FAR struct file *filep, FAR const char *relpath, int oflags, mode_t mode)
+{
+	FAR struct logsave_file_s *attr = (struct logsave_file_s *)kmm_zalloc(sizeof(struct logsave_file_s));
+
+	fvdbg("Open '%s'\n", relpath);
+	
+	filep->f_priv = (FAR void *)attr;
+	lldbg("%s open is called here\n", relpath);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: logsave_close
+ ****************************************************************************/
+
+static int logsave_close(FAR struct file *filep)
+{
+	return OK;
+}
+
+/****************************************************************************
+ * Name: logsave_read
+ ****************************************************************************/
+
+static ssize_t logsave_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	/* to be implemented according to network module requirements */
+	read_log_net();
```

Comment:
could you have "logsave_read" return an appropriate code if no logs are saved so that a line may be printed saying no logs were saved?

Reply:
Irrespective of whether logs were saved or not, the function can return OK as long as the underlying function does not encounter a memory allocation problem or something like that

Reply:
yes same opinion as @amogh-samsung .
1. log should be copied to user buffer, as I know goal is that send log to some kind of server, am I right?
2. Then you should maintain current position of buffer, because user will use kind of below example.
```
int fd = open("/proc/log", O_RDONLY);
int ret;
while (ret > 0) {
     ret = read(fd, buf, sizeof(buf));
}
close(fd);
```
3. If new logs come to uart buffer during user's reading, what happened?
   For exmaple, buffer data is full, and user start to read first chunk now. But much of new logs comes to uart.
   In this case, latest log will be saved to first chunk, then user's log will be corrupted, I think.
   To prevent this, I think you should use ring buffer. How do you think?

Reply:
Hi @Taejun-Kwon , I didn't understand how ring buffer would prevent this as even in the case of ring buffer the contents get overwritten. Could you please correct me if I am wrong? The current implementation using the queue is similar to ring buffer as when the size is at max, we reuse the chunks from the head. And regarding the logs corruption, I think we should disable pre-emption during reading (like using irqsave()) so that a read is fully complete and there are no new logs added in between. Could you please let me know your thoughts regarding this